File: test/console.js
"use strict";
// JAB-002: the `console` object — log/info/debug → stdout (fd 1), warn/error →
// stderr (fd 2), Node-style multi-arg + %s/%d/%i/%f/%j/%o/%% formatting.  We
// intercept the io.writeAll leaf console layers over (buf.cpp) to capture both
// the fd and the exact formatted bytes in-process — so one .js test asserts
// routing AND formatting.  Each assertion throws -> non-zero exit (CTest fails).
function fail(m) { throw "FAIL " + m; }
function eq(a, b, m) { if (a !== b) fail(m + ": " + JSON.stringify(a) + " !== " + JSON.stringify(b)); }

let cap;
const real = io.writeAll;
io.writeAll = (fd, bytes) => { cap.push([fd, utf8.Decode(bytes)]); return bytes.length; };
function grab(fn) { cap = []; fn(); return cap; }
function one(fn) { const c = grab(fn); if (c.length !== 1) fail("expected 1 write, got " + c.length); return c[0]; }

try {
  // fd routing: log/info/debug -> stdout (1), warn/error -> stderr (2)
  eq(one(() => console.log("x"))[0], 1, "log->stdout");
  eq(one(() => console.info("x"))[0], 1, "info->stdout");
  eq(one(() => console.debug("x"))[0], 1, "debug->stdout");
  eq(one(() => console.warn("x"))[0], 2, "warn->stderr");
  eq(one(() => console.error("x"))[0], 2, "error->stderr");

  // trailing newline + plain text
  eq(one(() => console.log("hi"))[1], "hi\n", "newline");

  // multi-arg space join
  eq(one(() => console.log("a", "b", "c"))[1], "a b c\n", "multi-arg join");

  // printf specifiers
  eq(one(() => console.log("%s/%d", "x", 5))[1], "x/5\n", "%s %d");
  eq(one(() => console.log("%i", 3.9))[1], "3\n", "%i truncates");
  eq(one(() => console.log("%f", 3.5))[1], "3.5\n", "%f");
  eq(one(() => console.log("100%%"))[1], "100%\n", "%% literal");
  eq(one(() => console.log("50% off"))[1], "50% off\n", "bare % is literal");

  // extra args beyond specifiers are space-appended; missing args stay literal
  eq(one(() => console.log("%s", "a", "b"))[1], "a b\n", "extra args appended");
  eq(one(() => console.log("%s %s", "a"))[1], "a %s\n", "missing arg literal");

  // object / array inspection (%o and bare)
  eq(one(() => console.log("%o", { a: 1 }))[1], "{ a: 1 }\n", "%o object");
  eq(one(() => console.log({ a: 1, b: "c" }))[1], "{ a: 1, b: 'c' }\n", "bare object");
  eq(one(() => console.log([1, 2, 3]))[1], "[ 1, 2, 3 ]\n", "array");
  eq(one(() => console.log({}))[1], "{}\n", "empty object");

  // %j is JSON; a cycle degrades to [Circular] (no throw)
  eq(one(() => console.log("%j", { a: 1 }))[1], '{"a":1}\n', "%j json");
  const cyc = {}; cyc.self = cyc;
  eq(one(() => console.log("%j", cyc))[1], "[Circular]\n", "%j cycle");
  if (one(() => console.log(cyc))[1].indexOf("Circular") < 0) fail("bare cycle not handled");

  // non-string scalars stringified
  eq(one(() => console.log(42))[1], "42\n", "number");
  eq(one(() => console.log(true, null))[1], "true null\n", "bool+null");
  if (one(() => console.log(function foo() {}))[1].indexOf("Function") < 0) fail("function not inspected");

  // assert: truthy is silent, falsy writes "Assertion failed" to stderr
  eq(grab(() => console.assert(true, "nope")).length, 0, "assert(true) silent");
  const a = one(() => console.assert(false, "boom"));
  eq(a[0], 2, "assert->stderr");
  if (a[1].indexOf("Assertion failed") < 0) fail("assert text: " + a[1]);
} finally {
  io.writeAll = real;
}

io.log("console.js OK");

File: console.cpp
#include "JABC.hpp"

//  JAB-002: a standard `console` — pure JS (like ansi.cpp / buf.cpp), layering
//  over the existing leaves, no native encoding or raw IO of its own.  Output
//  goes through `utf8.Encode` + `io.writeAll` to a chosen fd: log/info/debug →
//  stdout (1), warn/error/trace/assert → stderr (2), like Node.  `format` does
//  the multi-arg space-join + `%s %d %i %f %j %o %O %c %%` expansion; `inspect`
//  is a small JSON-ish stringifier with a cycle guard (`[Circular]`).
static const char* JABC_CONSOLE_JS = R"JS(
(function (g) {
  "use strict";
  const io = g.io, utf8 = g.utf8;

  //  A small util.inspect: scalars stringified, objects/arrays rendered
  //  JSON-ish (single-quoted strings, bareword keys), cycles → [Circular].
  function inspect(v, seen) {
    const t = typeof v;
    if (t === "string") return "'" + v.replace(/\\/g, "\\\\")
        .replace(/'/g, "\\'").replace(/\n/g, "\\n") + "'";
    if (t === "bigint") return String(v) + "n";
    if (t === "function") return "[Function" + (v.name ? ": " + v.name : " (anonymous)") + "]";
    if (t === "symbol") return v.toString();
    if (t !== "object" || v === null) return String(v);  // number/boolean/undefined/null
    seen = seen || [];
    if (seen.indexOf(v) >= 0) return "[Circular]";
    if (v instanceof Error) return v.stack || (v.name + ": " + v.message);
    seen.push(v);
    let out;
    if (Array.isArray(v)) {
      out = v.length ? "[ " + v.map((x) => inspect(x, seen)).join(", ") + " ]" : "[]";
    } else {
      const keys = Object.keys(v);
      const body = keys.map((k) => {
        const key = /^[A-Za-z_$][\w$]*$/.test(k) ? k : "'" + k + "'";
        return key + ": " + inspect(v[k], seen);
      });
      out = body.length ? "{ " + body.join(", ") + " }" : "{}";
    }
    seen.pop();
    return out;
  }

  //  Stringify one non-specifier arg: a top-level string prints raw (no
  //  quotes), everything else goes through inspect.
  function str(a) { return typeof a === "string" ? a : inspect(a); }

  //  Node-style util.format over an arguments/array `a`.
  function format(a) {
    if (a.length === 0) return "";
    let i = 1, out;
    const first = a[0];
    if (typeof first === "string" && /%[sdifjoOc%]/.test(first)) {
      out = first.replace(/%([sdifjoOc%])/g, (m, k) => {
        if (k === "%") return "%";
        if (i >= a.length) return m;          // not enough args: leave literal
        const x = a[i++];
        switch (k) {
          case "s": return typeof x === "object" && x !== null ? inspect(x)
              : (typeof x === "bigint" ? String(x) + "n" : String(x));
          case "d": case "i": {
            if (typeof x === "bigint") return String(x) + "n";
            const n = Number(x); return Number.isNaN(n) ? "NaN" : String(Math.trunc(n));
          }
          case "f": { const n = Number(x); return Number.isNaN(n) ? "NaN" : String(n); }
          case "j": try { return JSON.stringify(x); } catch (e) { return "[Circular]"; }
          case "o": case "O": return inspect(x);
          case "c": return "";                // CSS directive: consumed, no output
        }
      });
    } else {
      out = str(first);
    }
    for (; i < a.length; i++) out += " " + str(a[i]);
    return out;
  }

  //  The one write path: format → utf8 bytes → io.writeAll to the fd (reusing
  //  the buf.cpp leaves; routing through io.writeAll keeps it interceptable).
  function emit(fd, s) { io.writeAll(fd, utf8.Encode(s + "\n")); }

  g.console = {
    log:   function () { emit(io.stdout, format(arguments)); },
    info:  function () { emit(io.stdout, format(arguments)); },
    debug: function () { emit(io.stdout, format(arguments)); },
    warn:  function () { emit(io.stderr, format(arguments)); },
    error: function () { emit(io.stderr, format(arguments)); },
    trace: function () { emit(io.stderr, "Trace: " + format(arguments)); },
    dir:   function (obj) { emit(io.stdout, inspect(obj)); },
    assert: function (cond) {
      if (cond) return;
      const rest = Array.prototype.slice.call(arguments, 1);
      emit(io.stderr, rest.length ? "Assertion failed: " + format(rest) : "Assertion failed");
    },
  };
})(this);
)JS";

ok64 JABCConsoleInstall() {
  JABCExecute(JABC_CONSOLE_JS);
  return OK;
}
